File: Sliding Window Pattern/Longest_Repeating_Character_Replacement.cpp
#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;

int longestRepeatingCharacterReplacement(string s, int k) {
    int windowStart = 0, maxLength = 0, maxRepeatLetter = 0;
    unordered_map<char, int> dict;
    for (int windowEnd = 0; windowEnd < s.length(); windowEnd++) {
        char currChar = s[windowEnd];
        dict[currChar]++;
        maxRepeatLetter = max(maxRepeatLetter, dict[currChar]);
        if ((windowEnd - windowStart + 1 - maxRepeatLetter) > k) {
            char leftChar = s[windowStart];
            dict[leftChar]--;
            windowStart++;
        }
        maxLength = max(maxLength, windowEnd - windowStart + 1);
    }
    return maxLength;
}

int main() {
    string s = "aabccbb";
    int k = 2;
    cout<<longestRepeatingCharacterReplacement(s, k)<<endl;
    return 0;
}

File: Sliding Window Pattern/Longest_Repeating_Character_Replacement.py
def longest_repeating_character_replacement(s:str, k:int):
    windowStart = 0
    maxLength = 0
    maxRepeatLetter = 0
    dictionary = {}
    for windowEnd in range(len(s)):
        currChar = s[windowEnd]
        dictionary[currChar] = 1 + dictionary.get(currChar, 0)
        maxRepeatLetter = max(maxRepeatLetter, dictionary[currChar])
        if (windowEnd - windowStart + 1 - maxRepeatLetter) > k:
            leftChar = s[windowStart]
            dictionary[leftChar] -= 1
            windowStart += 1
        maxLength = max(maxLength, windowEnd - windowStart + 1)
    return maxLength

if __name__ == "__main__":
    s = "aabccbb"
    k = 2
    print(longest_repeating_character_replacement(s, k))